### PR TITLE
Fix bug: Async client may not close

### PR DESCRIPTION
**Pull request by jokemanfire**
when just new a Client,the fd will increase.
I also think the lifetime in async client is too long. also in sync #225 .

I obtained the following test code for this phenomenon：
```
fn get_describe_numbers(pid: &str) -> i64 {
    let proc_dir = format!("/proc/{}/fd", pid);
    let mut cnt = 0;
    match read_dir(PathBuf::from(proc_dir)) {
        Ok(entries) => {
            for entry in entries {
                match entry {
                    Ok(e) => {
                        cnt += 1;
                    }
                    Err(e) => {
                        eprintln!("Error reading directory entry: {}", e);
                    }
                }
            }
        }
        Err(e) => {
            eprintln!("Error reading directory: {}", e);
        }
    }
    cnt
}

#[cfg(unix)]
#[tokio::main(flavor = "current_thread")]
async fn main() {
    let pid = process::id().to_string();
    let pre_fds = get_describe_numbers(pid.as_str());
    for i in 0..10 {
        let now_fds = get_describe_numbers(pid.as_str());
        if now_fds != pre_fds {
            println!("fd is not release pre_fd {pre_fds:?}  now_fd {now_fds:?}");
            // pre_fds = now_fds
        }
        println!(
            "-------------now fd number is {now_fds:?}---------------------------------------"
        );
        let c = Client::connect(utils::SOCK_ADDR).unwrap();
    }
    let now_fds = get_describe_numbers(pid.as_str());
    if now_fds != pre_fds {
        println!("fd is not release pre_fd {pre_fds:?}  now_fd {now_fds:?}");
        // pre_fds = now_fds
    }

    println!("over");
}
```

### CODE REVIEW

**Comment by jokemanfire** [comment]
The Delegate thread should be release , I missed that.

File: src/asynchronous/client.rs [hunk 1]
```diff
@@ -45,6 +46,7 @@ impl Client {
     pub fn new(fd: RawFd) -> Client {
         let stream = utils::new_unix_stream_from_raw_fd(fd);
```
Comment (Tim-Zhang) [inline src/asynchronous/client.rs:47]: The fd will be closed when ClientClose dropping

Comment (jokemanfire) [reply src/asynchronous/client.rs:47]: hi, It looks like async client didn't hold ClientClose? If I miss?

**Review by wllenyj — APPROVED** [review]

**Review by wllenyj — COMMENTED** [review]
The `Connection` holds the fd. Did the fd close when conn was dropped?
```
        let conn = Connection::new(stream, delegate);
        tokio::spawn(async move { conn.run().await });
```

**Comment by jokemanfire** [comment]
> let conn = Connection::new(stream, delegate);

I think connection is client's member, the lifetime of connection should not above the client struct, if client is drop as well as conn should exit.

**Comment by wllenyj** [comment]
I am not sure, is not the fd closed by the Connection. If true, there is now a risk of a double close.
If any fd is opened between two closes, it will be closed silently.

**Comment by jokemanfire** [comment]
> I am not sure, is not the fd closed by the Connection. If true, there is now a risk of a double close. If any fd is opened between two closes, it will be closed silently.

I check the connection  , but there's no fd rlease, please check it again if i missed.

**Comment by jokemanfire** [comment]
Raw fd looks like  difficulty to  manage ownership and lifecycle  how about use this https://docs.rs/tokio/latest/tokio/net/struct.UnixStream.html. If it is reasonable , I'd like to change it . @wllenyj @Tim-Zhang

**Comment by Tim-Zhang** [comment]
Got it. It's not a problem because the fd will be closed in https://github.com/containerd/ttrpc-rust/blob/d2a0ce1fa10297b9b4e7b883ab456736e1d6f209/src/asynchronous/connection.rs#L72 by design. But there is a problem with the test code is that it exited too quickly to start the tokio spawn. Try to add `tokio::time::sleep(Duration::from_secs(1)).await;` after the for loop. @jokemanfire

**Comment by jokemanfire** [comment]
> Got it. It's not a problem because the fd will be closed in
> 
> https://github.com/containerd/ttrpc-rust/blob/d2a0ce1fa10297b9b4e7b883ab456736e1d6f209/src/asynchronous/connection.rs#L72
> 
> by design. But there is a problem with the test code is that it exited too quickly to start the tokio spawn. Try to add `tokio::time::sleep(Duration::from_secs(1)).await;` after the for loop. @jokemanfire

Thanks  I will close this pr. But in RFC https://rustmagazine.github.io/rust_magazine_2021/chapter_11/io-safety.html , we may should think about more safety IO like owendfd

**Pull request closed without merging**